src/libp2p/chain_exchange/provider.rs

```diff
@@ -21,68 +21,81 @@ pub fn make_chain_exchange_response<DB>(
 where
     DB: Blockstore + Send + Sync + 'static,
 {
-    let mut response_chain: Vec<TipsetBundle> = Vec::with_capacity(request.request_len as usize);
-
-    let mut curr_tipset_cids = request.start.clone();
-
-    loop {
-        let mut tipset_bundle: TipsetBundle = TipsetBundle::default();
-        let tipset = match cs.tipset_from_keys(&TipsetKeys::from_iter(curr_tipset_cids)) {
-            Ok(tipset) => tipset,
-            Err(err) => {
-                debug!("Cannot get tipset from keys: {}", err);
+    if !request.is_options_valid() {
+        return ChainExchangeResponse {
+            chain: Default::default(),
+            status: ChainExchangeResponseStatus::BadRequest,
+            message: format!("Invalid options {}", request.options),
+        };
+    }
 
+    let root = match cs.tipset_from_keys(&TipsetKeys::from_iter(request.start.clone())) {
+        Ok(tipset) => tipset,
+        Err(err) => {
+            debug!("Failed to get tipset from keys: {err}");
+            if let crate::chain::store::Error::NotFound(_) = err {
+                return ChainExchangeResponse {
+                    status: ChainExchangeResponseStatus::BlockNotFound,
+                    chain: Default::default(),
+                    message: "Start tipset was not found in the database".into(),
+                };
+            } else {
                 return ChainExchangeResponse {
                     chain: vec![],
                     status: ChainExchangeResponseStatus::InternalError,
-                    message: "Tipset was not found in the database".into(),
+                    message: err.to_string(),
                 };
             }
-        };
-
-        if request.include_messages() {
-            match compact_messages(cs.blockstore(), &tipset) {
-                Ok(compacted_messages) => tipset_bundle.messages = Some(compacted_messages),
-                Err(err) => {
-                    debug!("Cannot compact messages for tipset: {}", err);
-
-                    return ChainExchangeResponse {
-                        chain: vec![],
-                        status: ChainExchangeResponseStatus::InternalError,
-                        message: "Can not fulfil the request".into(),
-                    };
-                }
-            }
         }
+    };
 
-        curr_tipset_cids = tipset.parents().cids.clone().into_iter().collect();
-        let tipset_epoch = tipset.epoch();
-
-        if request.include_blocks() {
-            // Cloning blocks isn't ideal, this can maybe be switched to serialize this
-            // data in the function. This may not be possible without overriding rpc in
-            // libp2p
-            tipset_bundle.blocks = tipset.blocks().to_vec();
-        }
+    match cs
+        .chain_index
+        .chain(root)
+        .take(request.request_len as _)
+        .try_fold(
```

Comment:
Why fold instead of map?

Reply:
`try_fold` allows to terminate early on internal errors. 

> try_fold() takes two arguments: an initial value, and a closure with two arguments: an 'accumulator', and an element. The closure either returns successfully, with the value that the accumulator should have for the next iteration, or it returns failure, with an error value that is propagated back to the caller immediately (short-circuiting).

Reply:
Instead of manually dealing with errors, how about we split the function into two parts: An inner part that returns `Result<Vec<FullTipset>>` (or something similar) and an outer part that maps the result to `Success`, `PartialResponse`, or `InternalError`. That way, we don't have to return `InternalError` in multiple locations and we can use `?` as we normally would.

Reply:
I'm not sure if the outer part is a `map` function, as the function actually folds `Iterator<Item=Tipset>` into a single `ChainExchangeResponse`

> we don't have to return InternalError in multiple locations

Currently `InternalError` is returned in 2 places, 1 is to prepare the input of  `ChainIndex::chain`, the other is inside `try_fold`, I'm not sure I understand how to omit occurrence 1 with `map`

 could you elaborate on both points?